design/ceph/ceph-orchestrator-osd-add-remove.md

```diff
@@ -0,0 +1,147 @@
+2. Users of the Ceph Orchestrator CLI and/or Ceph Dashboard should be able to
+   create OSDs on nodes with as much complexity as `ceph-volume` is able to support all possible
+   user needs.
+3. Give as much room as possible for the Ceph Mgr Orchestrator Module and `ceph-volume` to implement
+   OSD management logic so that Rook's code can be simpler and long-term code maintenance on Rook
+   can be reduced.
+
+## Current status
+
+Rook expects to manage storage declaratively. Users can add disks to a Rook cluster by adding them
+to the `CephCluster` resource via the disk name itself, via 2 forms of pattern matching, or via the
+`useAllDisks` option. The Ceph Mgr Orchestrator Module differently expects to add disks to a Ceph
+cluster imperatively (not declaratively) by issuing a `ceph-volume` command with a Ceph
+[Drive Groups](https://github.com/ceph/ceph/blob/master/src/python-common/ceph/deployment/drive_group.py)
+specification. Once the `ceph-volume` command using the Drive Groups spec has been run, Ceph does
+not keep track of the Drive Groups specification used.
+
+Ceph expects the on-disk content of disks to be the declarative state of whether an OSD belongs to a
+Ceph cluster or not and what its configuration is. Rook is currently able to read this on-disk
+content via `ceph-volume` to determine if it should run a Kubernetes Deployment for a disk-based
+OSD.
+
+Rook currently does not implement OSD removal via any declarations in the `CephCluster` resource in
+order to prioritize the safety of user data. If a user wishes to remove an OSD, they must perform
+manual, imperative steps to do so.
+
+## Proposal
+
+The chief difficulty with this proposal is to harmonize interactions between Rook, which ideally
+wants to use Kubernetes manifests for a declarative intent for a Ceph cluster, and the Ceph
+Orchestrator Mgr Module, which treats OSD creation as imperative and allows the on-disk content of
+OSDs to be the declarative state.
+
+### Adding OSDs
+This design proposes to add a section to the `CephCluster` resource's `storage` configuration
+section for individual `nodes` to have disks added via a Drive Groups specification. The spec is
+proposed below with behavior described to follow.
+
+```yaml
+apiVersion: ceph.rook.io/v1
+kind: CephCluster
+spec:
+  storage:
+    nodes: # preexisting
+      - name: < node name > # preexisting
+        driveGroups: # proposed new
+          - removeFromSpecAfterCompmletion: < default false > # proposed new
+            spec: < JSON string to pass to ceph-volume for OSD creation >
+```
+
+Each individual `driveGroups` item's `spec` will be a simple string containing a JSON blob that gets
+passed along without modification or inspection by Rook to `ceph-volume` when running the OSD
+provisioning job on the corresponding node. The `ceph-volume` command to create OSDs via the Drive
+Groups spec is given priority and executed first before any other Rook-based OSD creation methods
```

Comment:
If a driveGroup is specified for a node, seems like we should skip all other OSD creation for that node. This means the useAllDevices and deviceFilter would be ignored in that case. Alternatively, maybe we should still apply the other OSD creation techniques, but just document the recommendation to set `useAllDevices: false` and no   `deviceFilter` if they are specifying driveGroups. It could get confusing quickly if multiple methods are used.

Reply:
I think that an "automatic-provision" mode should be created.
In the automatic provision mode the user can decide if he is going to use "drive groups" or the clasic attributes (useAllDevices...) (mutually exclusive)
If the "automatic provision" mode is not set , the user should include the OSD's items in the Cluster CRD, or use the dashboard/orchestratror-cli to create the OSD's using drive groups.

Reply:
That is not the intent, @travisn. It is merely that the Drive Groups are executed first--are executed before provisioning based on device filters or useAllDevices..

Reply:
Ok, in the docs for driveGroups how about if we document the recommendation to not set device filters or useAllDevices?

Reply:
fwiw DriveGroups also have a `useAllDevices` equivalent iirc. @jschmid1 ?

Reply:
right, the `all: true` option

Reply:
> Ok, in the docs for driveGroups how about if we document the recommendation to not set device filters or useAllDevices?

The need for this documentation is included in the design.